File: src/clj/pdp/core/naive.clj
;; Copyright (c) Konrad Grzanek. All rights reserved.
;; Created 2015-11-19

(in-ns 'pdp.core)

;; BASED ON LINKED LIST

(deftype LinkedEvaluator [rules])

(defmacro def-linked-evaluator
  []
  `(def ~'current-ev (pdp.core.LinkedEvaluator. (java.util.LinkedList.))))


(extend-type LinkedEvaluator
  Evaluator
  (add-rule [this rule]
    (synchronized
     this
     (.addLast ^java.util.LinkedList (.rules this) rule)))

  (eval-conf-impl [this conf]
    ;; Intentionally we go with no synchronization, relying on the
    ;; assumption that the process of creating (and thus adding) rules
    ;; is completed before any evaluation.
    (jpdp.core.Naive/evalLinked (.rules this) conf)))


;; BASED ON ARRAY LIST

(deftype ArrayEvaluator [rules])

(defmacro def-array-evaluator
  []
  `(def ~'current-ev (pdp.core.ArrayEvaluator. (java.util.ArrayList.))))


(extend-type ArrayEvaluator
  Evaluator
  (add-rule [this rule]
    (synchronized
     this
     (.add ^java.util.ArrayList (.rules this) rule)))

  (eval-conf-impl [this conf]
    ;; See comment for LinkedEvaluator's implementation.
    (jpdp.core.Naive/evalArray (.rules this) conf)))

File: src/java/jpdp/core/Naive.java
/*
 * Copyright (c) Konrad Grzanek. All rights reserved.
 * Created 2015-11-19
 */
package jpdp.core;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;

import clojure.lang.RT;

public final class Naive {

  public static boolean evalLinked(LinkedList<Rule> rules, Object conf) {
    final Iterator<Rule> it = rules.iterator();
    while (it.hasNext()) {
      final Rule rule = it.next();
      if (!RT.booleanCast(rule.f.invoke(conf))) {
        return false;
      }
    }
    return true;
  }

  public static boolean evalArray(ArrayList<Rule> rules, Object conf) {
    for (int i = 0, n = rules.size(); i < n; i++) {
      final Rule rule = rules.get(i);
      if (!RT.booleanCast(rule.f.invoke(conf))) {
        return false;
      }
    }
    return true;
  }

  private Naive() {

  }

}
